<div style="text-align: center; padding: 20px; font-family: Vazir;">
<h1 align="center" style="font-size: 28px; color:rgb(64, 244, 202); width: 100%;">⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️<br>تمرین شماره ۵ بخش دوم <br>⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️</h1>
<h2 align="center" style="color:rgb(90, 255, 184); font-size: 20px;">تراول‌بات: چت‌بات هوشمند برنامه‌ریزی سفر</h2>
<p align="center" style="color: #666; font-size: 16px;">طراح تمرین: عرفان شهابی</p>
<p align="center" style="color: #666; font-size: 16px; margin-bottom: 30px;">erfanshahabi@outlook.com</p>

<div dir="rtl" style="border: 2px dashed rgb(90, 255, 184); border-radius: 8px; padding: 20px; margin: 20px auto; max-width: 500px; text-align: right;">
<p style="color: rgb(64, 244, 202); font-size: 18px; margin-bottom: 15px;">📝 مشخصات دانشجو:</p>
<p style="color: #666; margin: 5px;">نام و نام خانوادگی: [نام و نام خانوادگی خود را وارد کنید]</p>
<p style="color: #666; margin: 5px;">شماره دانشجویی: [شماره دانشجویی خود را وارد کنید]</p>
<p style="color: #666; margin: 5px;">تاریخ ارسال: [تاریخ ارسال را وارد کنید]</p>
</div>
</div>

<div dir="rtl" style="text-align: justify; padding: 25px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
<div style="line-height: 2.0; font-size: 17px; color: black; font-family: Vazir;">
<div style="padding-right:40px">
در این تمرین، شما به طراحی و پیاده‌سازی یک چت‌بات به نام <b>تراول‌بات</b> خواهید پرداخت. تراول‌بات بخشی از یک آژانس مسافرتی آنلاین فرضی است که هدف آن ارائه خدمات به کاربران از جمله پیدا کردن بلیط پرواز، جستجوی هتل، پیشنهاد رستوران، ارائه اطلاعات از مقصد و برنامه‌ریزی روزانه برای مسافران است.

در این پروژه، شما با استفاده از فریم‌ورک <b>LangGraph</b> یک چت‌بات هوشمند طراحی خواهید کرد که قابلیت پاسخگویی به سوالات کاربران و اجرای سرویس‌های مرتبط با دستیار مسافرتی را داشته باشد. این تمرین فرصتی است تا شما به صورت عملی با فرآیند طراحی و توسعه چت‌بات‌ها آشنا شوید.

در فرآیند این تمرین، شما علاوه بر طراحی چت‌بات، با ابزارهایی نظیر:
- <b>LanceDB</b>: برای ذخیره‌سازی و بازیابی هوشمند اطلاعات با RAG
- <b>Amadeus API</b>: برای جستجوی پرواز و هتل با قیمت‌های واقعی
- <b>Tavily Search</b>: برای جستجوی وب و اطلاعات به‌روز
- <b>"BAAI/bge-small-en-v1.5"</b>: برای embedding و جستجوی معنایی

کار خواهید کرد. برای هسته پردازش زبان طبیعی چت‌بات خود می‌توانید با انتخاب خودتان از بین مدل‌های زبانی موجود مانند <b>Gemini 3 Flash</b>, <b>Claude 4.5 Sonnet</b>, <b>GPT-5</b>, <b>o1-mini</b>، یا <b>Llama 3.3 70B</b> استفاده کنید که API آن‌ها از طریق وب‌سایت ارائه‌دهندگان در دسترس است.

<b>توجه:</b> با توجه به عدم امکان دسترسی رایگان به مدل‌های قوی‌تر در زبان فارسی، زبان ارتباط با چت‌بات در تمامی مراحل <b>انگلیسی</b> خواهد بود.

طی هر مرحله از تمرین، شما چت‌بات خود را تست و ارزیابی خواهید کرد تا اطمینان حاصل کنید که خروجی طراحی شده از نظر کارایی و تعامل‌پذیری، استانداردهای لازم را دارا باشد.
</div>
<br>
<div style="padding-right:100px">
📋 <b>ساختار تمرین:</b>
<br><br>
<b>بخش اول: پیکربندی محیط (۲.۵ نمره)</b>
<br>
<b>بخش دوم: آماده‌سازی مجموعه داده (۲.۵ نمره)</b>
<br>
<b>بخش سوم: ساخت پایگاه داده برداری (۷.۵ نمره)</b>
<br>
<b>بخش چهارم: تعریف ابزارها (۲۰ نمره)</b>
<br>
<b>بخش پنجم: ساخت Agent با LangGraph (۱۲.۵ نمره)</b>
<br>
<b>بخش ششم: تست Agent با سناریوهای مختلف (۱۰ نمره)</b>
<br><br>
<b>⭐ مجموع نمرات بخش دوم تمرین پنج: ۵۵ نمره</b>
<br>
</div>
</div>
<div dir='rtl' style="line-height: 1.8; font-family: Vazir; font-size: 16px; margin-top: 20px; background-color: #e8eaf6; padding: 15px; border-radius: 8px; color:black">
💡 <b>نکات مهم:</b>
<br>
- تمام کدها باید کامل اجرا شده و خروجی داشته باشند<br>
- از API Key های معتبر و فعال استفاده کنید<br>
- به توضیحات تشریحی و تحلیل نتایج توجه ویژه کنید<br>
- از فونت <b>Vazir</b> و راست‌چین برای متن‌های فارسی استفاده کنید<br>
- تمام پاسخ‌های چت‌بات باید به <b>انگلیسی</b> باشد<br>
- فایل نهایی را با فرمت <code>NLP_CA5_LASTNAME_STUDENTID.ipynb</code> نامگذاری کنید<br>
</div>
</div>

# <div style="text-align: center; direction: rtl; font-family: Vazir;"><h1 align="center" style="font-size: 24px; padding: 20px;">⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️<br>راه اندازی و پیکر بندی (۲.۵نمره)<br>⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️</h1></div>


<p dir="rtl" style="text-align: right; padding:30px; background-color:rgb(12, 12, 12); border-radius: 12px; color: white; font-family: Vazir;">
در این بخش، محیط کاری خود را برای ساخت تراول‌بات آماده می‌کنیم. شما باید کتابخانه‌های مورد نیاز را نصب و API Keyهای لازم را تنظیم کنید.

<p dir="rtl" style="padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
در این بخش با سه سرویس API کار خواهیم کرد:

• <b>Language Model API</b>: برای مدل زبانی هوشمند<br>
• <b>Tavily API</b>: برای جستجوی وب<br>
• <b>Amadeus API</b>: برای جستجوی پروازها و هتل‌ها<br>
</p>
</p>


## <div style="text-align: center; direction: rtl; font-family: Vazir;">نصب کتابخانه‌ها</div>


<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">

کتابخانه‌های مورد نیاز را که در فایل requirements.txt در پوشه تمرین قرار داده شده است را نصب و فراخوانی کنید.
</div>


In order to be able to do this we first need to upload the requirements.txt files on google drive, after that we can install them like so:

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

!pip install -r /content/drive/MyDrive/NLP_CA5_Q2/requirements.txt

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


Now we test them.

In [ ]:
import langgraph
import langchain_google_genai
import tavily
import langchain_core
import amadeus
import lancedb
import sentence_transformers
import llama_parse
import llama_index
import kagglehub
import llama_index.core
import llama_index.readers.file
import dotenv
import pandas
import langchain_openai
import langchain_community
import Levenshtein
import langchain_text_splitters
import langchain

## <div style="text-align: center; direction: rtl; font-family: Vazir;">API Keys</div>


<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
<p style="line-height: 1.8; text-align: right;">
در این مرحله باید API Keyهای مورد نیاز را تنظیم نمایید.
همانطور که گفته شد در این تمرین به سه API Key نیاز خواهیم داشت:

• <b>Language Model API</b>: برای مدل زبانی هوشمند<br>
• <b>Tavily API</b>: برای جستجوی وب<br>
• <b>Amadeus API</b>: برای جستجوی پروازها و هتل‌ها<br>
</p>
</div>


The most secure way to use APIs is to set them as environment variables from a terminal (or google colab) but since this notebook might be run by other people I've added the enviornment variables from the code itself

In [ ]:
import os

os.environ["GOOGLE_API_KEY"] = "AIzaSyDZKSVCzrxMTXuW23-sZ9T7kMPW8_jJ8BE"
os.environ["TAVILY_API_KEY"] = "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"
os.environ["AMADEUS_CLIENT_ID"] = "Nfz3xehfTLvmODGuWyd6CAgFDkrjfAhr"
os.environ["AMADEUS_CLIENT_SECRET"] = "EL7oGOCJjfU2TjLE"

# <div style="text-align: center; direction: rtl; font-family: Vazir;"><h1 align="center" style="font-size: 24px; padding: 20px;">⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️<br>مجموعه دادگان (۲.۵نمره)<br>⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️</h1></div>


<p dir="rtl" style="text-align: right; padding:30px; background-color:rgb(12, 12, 12); border-radius: 12px; color: white; font-family: Vazir;">
در این بخش، مجموعه‌ دادگان مورد نیاز را بارگزاری و پردازش می‌کنیم.
<p dir="rtl" style="padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
در این تمرین ما سه مجموعه داده خواهیم داشت:

• <b>zinovadr/iata-airport-code</b>:مربوط به کد اختصاصی فرودگاه‌های جهان<br>
• <b>"phanee16/currency-and-country-code-mapping"</b>:مربوط به کد اختصاصی ارز کشورهای جهان<br>
• <b>FAQ</b>: پرسش و پاسخ‌های پرتکرار آژانس های مسافرتی<br>
• <b>World Travel Book</b>:کتاب جهانگردی شامل مکان های توریستی و دیدنی شهرهای جهان<br>



## <div style="text-align: center; direction: rtl; font-family: Vazir;">Download Kaggle IATA Dataset</div>


<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">

در این بخش لازم است با استفاده از ابزار مناسب مجموعه داده (zinovadr/iata-airport-code) را از بستر Kaggle دانلود نموده و یک Dictionary mapping هوشمند از نام شهر به کد فرودگاه مورد نظر بسازید.

</div>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
• تعداد فرودگاه‌های دارای کد IATA<br>
• تعداد ورودی‌های mapping dictionary<br>
• نمونه جستجو: Dubai → DXB، Tehran → THR
</p>

First let's load the dataset.

In [ ]:
import kagglehub

path = kagglehub.dataset_download("zinovadr/iata-airport-code")

print("Path to dataset files:", path)

df = pandas.read_csv(os.path.join(path, "airport-codes_csv.csv"))

Using Colab cache for faster access to the 'iata-airport-code' dataset.
Path to dataset files: /kaggle/input/iata-airport-code


Now let's take a look at the dataset

In [ ]:
df.head()

,ident,type,name,elevation_ft,continent,iso_country,iso_region,municipality,gps_code,iata_code,local_code,coordinates
0,00A,heliport,Total Rf Heliport,11.0,NaN,US,US-PA,Bensalem,00A,NaN,00A,"-74.93360137939453, 40.07080078125"
1,00AA,small_airport,Aero B Ranch Airport,3435.0,NaN,US,US-KS,Leoti,00AA,NaN,00AA,"-101.473911, 38.704022"
2,00AK,small_airport,Lowell Field,450.0,NaN,US,US-AK,Anchor Point,00AK,NaN,00AK,"-151.695999146, 59.94919968"
3,00AL,small_airport,Epps Airpark,820.0,NaN,US,US-AL,Harvest,00AL,NaN,00AL,"-86.77030181884766, 34.86479949951172"
4,00AR,closed,Newport Hospital & Clinic Heliport,237.0,NaN,US,US-AR,Newport,NaN,NaN,NaN,"-91.254898, 35.6087"


Based on this we can see that the column named 'muicipality' contains the city names and the column named 'iata_code' has the codes of the airport code so we will be working with these two columns for creating the mapping.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 55539 entries, 0 to 55538
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   ident         55539 non-null  object 
 1   type          55539 non-null  object 
 2   name          55539 non-null  object 
 3   elevation_ft  48502 non-null  float64
 4   continent     27630 non-null  object 
 5   iso_country   55292 non-null  object 
 6   iso_region    55539 non-null  object 
 7   municipality  49803 non-null  object 
 8   gps_code      41133 non-null  object 
 9   iata_code     9199 non-null   object 
 10  local_code    29012 non-null  object 
 11  coordinates   55539 non-null  object 
dtypes: float64(1), object(11)
memory usage: 5.1+ MB


In [ ]:
df = df[df["iata_code"].notna()]
print("Number of non-null airport codes:", df["iata_code"].notna().sum())

Number of non-null airport codes: 9199


However as we can see there are quite a few missing codes, in fact there are only '9,199' non-null airport codes. So we will only be using these for making the mapping.

In [ ]:
city_airport_mapping = {}
for index, row in df.iterrows():
    city = row["municipality"]
    airport_code = row["iata_code"]
    if city not in city_airport_mapping:
        city_airport_mapping[city] = []
    city_airport_mapping[city].append(airport_code)

## <div style="text-align: center; direction: rtl; font-family: Vazir;">Download Kaggle Currency Dataset</div>


<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">

در این بخش لازم است با استفاده از ابزار مناسب مجموعه داده ("phanee16/currency-and-country-code-mapping") را از بستر Kaggle دانلود نموده و یک Dictionary mapping هوشمند از نام کشور به کد ارز مورد نظر بسازید.

</div>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
• تعداد ارزها<br>
• تعداد ورودی‌های mapping dictionary<br>
• نمونه جستجو: IRAN → IRR, UAE → AED
</p>

In [ ]:
# WRITE YOUR CODE HERE


In [ ]:
# WRITE YOUR CODE HERE


## <div style="text-align: center; direction: rtl; font-family: Vazir;">FAQ</div>

<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
مجموعه داده پرسش و پاسخ‌های پرتکرار یک آژانس مسافرتی در فایل FAQ.js جهت استفاده در مراحل بعدی ساخت تراول‌بات در اختیار شما قرار گرفته است. این فایل داده را بارگزاری نمایید.
</div>

In [ ]:
# WRITE YOUR CODE HERE


## <div style="text-align: center; direction: rtl; font-family: Vazir;">parse Book</div>

<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">

در این قسمت از تمرین با استفاده از ابزار Llama Parse و API Key مخصوص به این ابزار، کتاب جهانگردی موجود در فایل تمرین را به Text تبدیل کرده و همراه با متادیتای آن در یک Dataframe ذخیره کنید

</div>


In [ ]:
# WRITE YOUR CODE HERE

<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
نمایش نمونه از داده متنی
</p>

# <div style="text-align: center; direction: rtl; font-family: Vazir;"><h1 align="center" style="font-size: 24px; padding: 20px;">⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️<br>ساخت پایگاه داده برداری‌(۷.۵ نمره)<br>⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️</h1></div>


<p dir="rtl" style="text-align: right; padding:30px; background-color:rgb(12, 12, 12); border-radius: 12px; color: white; font-family: Vazir;">
در این بخش از تمرین نیاز است دو Schema مناسب جهت ذخیره سوال و جواب‌های پر تکرار و متن کتاب جهانگردی مناسب برای LanceDB تعریف کنید و با استفاده از مدل امبدینگ "BAAI/bge-small-en-v1.5" سوال و جواب‌های پرتکرار و متن کتاب را به صورت جداگانه به امبدینگ تبدیل کرده و در جداول جداگانه ذخیره کنید.
<p dir="rtl" style="padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">

<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
• تبدیل سوال و جواب‌ها و متن کتاب به امبدینگ<br>
• تشکیل پایگاه داده برداری مبتنی بر LanceDB<br>
• نمونه جستجو موفق در پایگاه داده برداری
</p>

In [ ]:
# WRITE YOUR CODE HERE

# <div style="text-align: center; direction: rtl; font-family: Vazir;"><h1 align="center" style="font-size: 24px; padding: 20px;">⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️<br>تعریف ابزارها (۲۰ نمره)<br>⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️</h1></div>


<p dir="rtl" style="text-align: right; padding:30px; background-color:rgb(12, 12, 12); border-radius: 12px; color: white; font-family: Vazir;">
تراول بات چت‌ باتی است که قابلیت‌های مختلفی را دارا خواهد بود. در این قسمت نیاز است ابزارهای مختلفی را که به چت بات جهت انجام وظایف مورد انتظار کاربر کمک می‌کنند را تعریف کنید.
ابزارهای مورد نیاز به شرح زیر است:

• <b>Flight Search</b>

• <b>Hotel Search</b>

• <b>Resturant Search</b>

• <b>Get Weather Info</b>

• <b>Get Currency Info</b>

• <b>FAQ Search</b>

• <b>Trip Planning</b>




<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
• تعریف صحیح ابزارها<br>
• مدیریت خطاها و مقادیر ورودی از طرف کاربر<br>
• نمونه جستجو موفق توسط ابزارها
</p>

## <div style="text-align: center; direction: rtl; font-family: Vazir;">Flight Search</div>


<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
یکی از ابزارهای مهمی که تراول‌بات با استفاده از آن به کاربران خدمات می‌دهد، ابزار جستجوی پرواز است. شما باید ابزاری تعریف کنید که پارامترهایی مانند مبدا، مقصد و تاریخ پرواز را از کاربر دریافت کرده و با استفاده از Amadeus API به جستجوی پروازهای مورد نظر بپردازد. این ابزار باید حداقل ۳ پرواز را با اطلاعات کاملی شامل شماره پرواز، کد ایرلاین، قیمت دقیق (به همراه واحد پول)، تاریخ و ساعت حرکت، مدت پرواز و تعداد توقف‌ها به کاربر بازگرداند.

**نکات مهم پیاده‌سازی:**

- **تبدیل نام شهر به کد IATA:** ابزار Amadeus بر اساس کدهای یکتای فرودگاه (IATA Code) کار می‌کند. از تابع مپینگی که در بخش قبل نوشته‌اید استفاده کنید تا نام شهرها را به کد فرودگاه تبدیل کنید (مثلاً Tehran → IKA، Dubai → DXB).

- **مدیریت فرمت تاریخ:** Amadeus تاریخ را به فرمت YYYY-MM-DD می‌پذیرد. از آنجایی که ماهیت چت‌بات صحبت به زبان طبیعی است، کاربران فرمت تاریخ را به گونه‌های متفاوتی وارد می‌کنند(برای مثال به صورت متن و نه عدد). شما باید مدیریت فرمت تاریخ در ابزار و System Prompt را مد نظر داشته باشید.

- **مدیریت خطا:** در صورت بروز خطا در جستجو (مثل عدم یافتن پرواز یا خطای API)، پیام خطای مناسب و قابل فهم به کاربر بازگردانید.

- **استفاده از Decorator:** از `@tool` decorator برای تعریف این ابزار استفاده کنید تا LangGraph بتواند آن را شناسایی کند.

</div>


In [ ]:
# WRITE YOUR CODE HERE

## <div style="text-align: center; direction: rtl; font-family: Vazir;">Hotel Search</div>


<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
یکی دیگر از ابزارهای کاربردی تراول‌بات، جستجوی هتل است. شما باید ابزاری تعریف کنید که مقصد، تاریخ ورود (check-in) و تاریخ خروج (check-out) را از کاربر دریافت کرده و با استفاده از Amadeus API به یافتن هتل‌های مناسب در مقصد مورد نظر بپردازد. این ابزار باید لیستی از هتل‌های پیشنهادی را همراه با نام هتل، رتبه (rating)، قیمت شب، موقعیت مکانی به کاربر بازگرداند.

**نکات مهم پیاده‌سازی:**

- **استفاده از Amadeus Hotel Search API:** از endpoint مناسب Amadeus برای جستجوی هتل استفاده کنید. این API نیاز به کد شهر (IATA City Code) دارد که باید از mapping قبلی استفاده کنید.

- **تبدیل نام شهر به کد IATA:** مشابه جستجوی پرواز، از تابع از تابع مپینگ هوشمند برای تبدیل نام شهر به کد فرودگاه استفاده کنید (مثلاً Dubai → DXB).

- **فرمت تاریخ:** Amadeus تاریخ را به فرمت YYYY-MM-DD می‌پذیرد. از آنجایی که ماهیت چت‌بات صحبت به زبان طبیعی است، کاربران فرمت تاریخ را به گونه‌های متفاوتی وارد می‌کنند(برای مثال به صورت متن و نه عدد). شما باید مدیریت فرمت تاریخ در ابزار و System Prompt را مد نظر داشته باشید.

- **پارامتر Budget (اختیاری):** این ابزار می‌تواند یک پارامتر اختیاری budget داشته باشد که هتل‌ها را بر اساس محدوده قیمتی فیلتر کند.

- **خروجی کامل:** برای هر هتل، اطلاعاتی مانند نام، آدرس، رتبه و قیمت هر شب (به همراه واحد پول) را نمایش دهید.

- **مدیریت خطا:** در صورت عدم یافتن هتل یا خطای API، پیام مناسب به کاربر بازگردانید.

- **Fallback به Tavily:** در صورتی که Amadeus Hotel API در دسترس نبود یا محدودیت داشت، می‌توانید به عنوان fallback از Tavily Search استفاده کنید.

- **استفاده از Decorator:** از `@tool` decorator برای تعریف این ابزار استفاده کنید.
<div>


In [ ]:
# WRITE YOUR CODE HERE

## <div style="text-align: center; direction: rtl; font-family: Vazir;">Resturant Search</div>


<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
یکی دیگر از ابزارهای مفید تراول‌بات، جستجوی رستوران است. شما باید ابزاری تعریف کنید که نام مقصد را از کاربر دریافت کرده و با استفاده از Tavily Search API به جستجوی بهترین رستوران‌ها و غذاهای محلی در آن مقصد بپردازد. این ابزار باید لیستی از رستوران‌های پیشنهادی را همراه با نام، نوع غذا، توضیحات مختصر و لینک اطلاعات بیشتر به کاربر بازگرداند.

**نکات مهم پیاده‌سازی:**

- **استفاده از Tavily Search:** از آنجایی که اطلاعات رستوران‌ها و غذاهای محلی در وب به‌روز و متنوع هستند، از Tavily برای جستجوی این اطلاعات استفاده می‌کنیم.

- **ساخت Query مناسب:** query باید شامل کلمات کلیدی مناسب باشد. مثال: `"best restaurants local food in {destination}"` یا `"top rated restaurants cuisine in {destination}"`.

- **فقط پارامتر Destination:** برخلاف ابزار جستجوی پرواز و هتل، این ابزار فقط نام مقصد را به عنوان ورودی می‌گیرد.

- **خروجی متنوع:** نتایج باید شامل رستوران‌های مختلف با انواع غذاها باشد (غذای محلی، رستوران‌های معروف، کافه‌ها و ...).

- **فرمت خروجی:** هر رستوران باید شامل: نام، نوع غذا/آشپزی، توضیح کوتاه (100-150 کلمه)، و لینک URL برای اطلاعات بیشتر باشد.

- **محدود کردن نتایج:** تعداد نتایج را محدود کنید (مثلاً 5 رستوران برتر) تا خروجی خلاصه و کاربردی باشد.

- **مدیریت خطا:** در صورت عدم یافتن نتیجه یا خطای API، پیام واضح و مفیدی به کاربر بازگردانید.

- **استفاده از Decorator:** از `@tool` decorator برای تعریف این ابزار استفاده کنید.

**مثال Query:**
```
"best restaurants and local cuisine in Paris"
```
</div>

In [ ]:
# WRITE YOUR CODE HERE

## <div style="text-align: center; direction: rtl; font-family: Vazir;">Get Weather Info</div>


<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
ابزار دریافت اطلاعات آب و هوا یکی از ابزارهای حیاتی برای برنامه‌ریزی سفر است. شما باید ابزاری تعریف کنید که نام مقصد و تاریخ مورد نظر را از کاربر دریافت کرده و با استفاده از Tavily Search API اطلاعات پیش‌بینی آب و هوا را برای آن مقصد و تاریخ جستجو کند. این ابزار باید اطلاعاتی مانند دما، وضعیت هوا (آفتابی، بارانی، ابری)، رطوبت، احتمال بارندگی و توصیه‌های لباس‌پوشی را به کاربر بازگرداند.

**نکات مهم پیاده‌سازی:**

- **استفاده از Tavily Search:** برای دریافت پیش‌بینی آب و هوای به‌روز از جستجوی وب استفاده می‌کنیم، چون منابع هواشناسی آنلاین اطلاعات دقیق‌تری دارند.

- **ساخت Query مناسب:** query باید شامل نام مقصد و تاریخ باشد. مثال: `"weather forecast {destination} {date}"` یا `"weather in {destination} on {date}"`.

- **پارامترهای ورودی:**
  - `destination`: نام شهر یا مقصد (الزامی)
  - `date`: تاریخ مورد نظر به فرمت های مختلف

- **خروجی کامل:** اطلاعات آب و هوا باید شامل

- **پیش‌بینی چند روزه:** اگر ممکن است، اطلاعات پیش‌بینی چند روز آینده را نیز ارائه دهید.

- **مدیریت خطا:** در صورت عدم دسترسی به اطلاعات آب و هوا یا خطای API، پیام مناسب به کاربر بازگردانید.

- **استفاده از Decorator:** از `@tool` decorator برای تعریف این ابزار استفاده کنید.

**مثال Query:**
```
"weather forecast Dubai December 16 2025"
```
</div>

In [ ]:
# WITE YOUR CODE HERE

## <div style="text-align: center; direction: rtl; font-family: Vazir;">Get Currency Info</div>


<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
ابزار دریافت اطلاعات ارز یکی از ابزارهای ضروری برای مسافرانی است که به کشورهای مختلف سفر می‌کنند. شما باید ابزاری تعریف کنید که با توجه به نام کشور مبدا و مقصد و با استفاده از Tavily Search API نرخ تبدیل فعلی بین این دو ارز را جستجو و به کاربر بازگرداند. توجه داشته باشید از آنجایی که ماهیت چت‌بات صحبت به زبان طبیعی است، شما باید به درستی نام کشورها را به کد ارزی آن‌ها مپ کنید و در نهایت جست و جو را با کد ارزی انجام دهید. میتوانید مدیریت این قسمت را در ابزار و system Prompt انجام دهید.

**نکات مهم پیاده‌سازی:**

- **استفاده از Tavily Search:** برای دریافت نرخ تبدیل ارز به‌روز از جستجوی وب استفاده می‌کنیم تا آخرین نرخ‌های بازار را دریافت کنیم.

- **ساخت Query مناسب:** query باید شامل کدهای ارز باشد. مثال: `"USD to AED exchange rate"` یا `"EUR to JPY conversion rate today"`.

- **پارامترهای ورودی:**
  - نام کشور مبدا و مقصد

- **خروجی کامل:** اطلاعات باید شامل:
  - نرخ تبدیل فعلی (مثلاً: 1 USD = 3.67 AED)


- **پشتیبانی از کدهای ارز متداول:** ابزار باید کدهای استاندارد ISO 4217 را بپذیرد (USD، EUR، GBP، JPY، AED، CNY، INR و ...).

- **مدیریت خطا:**
  - شما باید از تابع mapping که نوشته اید برای map کردن نام هر کشور به کد ارزی مربوطه آن کشور استفاده کنید

- **استفاده از Decorator:** از `@tool` decorator برای تعریف این ابزار استفاده کنید.

**مثال Query:**
```
"United States Dollars to UAE Dirhams exchange rate today"
"Iranian Currency to France Currency conversion rate"
```

**مثال خروجی:**
```
1 USD = 3.67 AED
1 AED = 0.27 USD

```
</div>

In [ ]:
# WRITE YOUR CODE HERE

## <div style="text-align: center; direction: rtl; font-family: Vazir;">FAQ Search</div>


<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
ابزار جستجوی FAQ یکی از ابزارهای کلیدی برای پاسخ به سوالات متداول مسافران است. این ابزار به کاربران اجازه می‌دهد بدون نیاز به جستجوی وب، پاسخ سوالات رایج خود را درباره سفر دریافت کنند. شما باید ابزاری تعریف کنید که از سیستم RAG و LanceDB برای جستجوی معنایی در پایگاه دانش FAQ استفاده کند و مرتبط‌ترین پاسخ‌ها را به کاربر بازگرداند.

**نکات مهم پیاده‌سازی:**

- **استفاده از Vector Search:** از LanceDB و embedding های تشکیل شده برای جستجوی معنایی در پایگاه FAQ استفاده کنید. این روش سوالاتی که به شکل‌های مختلف پرسیده می‌شوند را تشخیص می‌دهد.

- **پارامتر ورودی:**
  - `query`: سوال کاربر به زبان طبیعی (الزامی)

- **محدود کردن نتایج:** از `limit(2)` استفاده کنید تا فقط 2 مرتبط‌ترین پرسش و پاسخ برگردانده شود و خروجی خلاصه باشد.

- **فرمت خروجی:** برای هر نتیجه:
  - سوال اصلی از پایگاه (Q:)
  - پاسخ کامل (A:)
  - فرمت واضح و خوانا


- **مدیریت نتایج خالی:** اگر هیچ FAQ مرتبطی یافت نشد، پیام `"No relevant FAQ found."` را برگردانید.

- **مدیریت خطا:** در صورت بروز خطا در جستجو، پیام خطا را به صورت واضح نمایش دهید.

- **استفاده از Decorator:** از `@tool` decorator برای تعریف این ابزار استفاده کنید تا LangGraph بتواند آن را شناسایی کند.

**مثال سوال کاربر:**
```
"What documents do I need for international travel?"
"Can I take food on the plane?"
"How early should I arrive at airport?"
```

**مثال خروجی:**
```
Based on our FAQ:

Q: What documents do I need for international travel?
A: For international travel, you typically need: 1) A valid passport (valid for at least 6 months), 2) Visa (if required for your destination), 3) Travel insurance, 4) Vaccination certificates (if required), 5) Return ticket.

Q: How do I get a visa?
A: Visa requirements depend on your destination and nationality. Apply through the embassy/consulate of your destination country...
```
</div>

In [ ]:
# WRITE YOUR CODE HERE

## <div style="text-align: center; direction: rtl; font-family: Vazir;">Trip Planning</div>


<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
ابزار برنامه‌ریزی سفر یکی از قدرتمندترین ابزارهای تراول‌بات است که با استفاده از سیستم RAG و پایگاه دانش راهنمای سفر، یک برنامه کامل و جامع برای سفر کاربر ایجاد می‌کند. شما باید ابزاری تعریف کنید که اطلاعات سفر را دریافت کرده و با جستجو در پایگاه دانش، یک itinerary روزانه همراه با پیشنهادات جاذبه‌ها، رستوران‌ها و فعالیت‌ها تولید کند.

**نکات مهم پیاده‌سازی:**

- **استفاده از RAG System:** از LanceDB و اطلاعاتی که از کتاب جهانگردی امبد کرده‌اید برای جستجوی اطلاعات مرتبط با مقصد استفاده کنید. چندین query مختلف بزنید تا اطلاعات جامعی جمع‌آوری شود.

- **پارامترهای ورودی:**
  - `destination`: نام شهر یا کشور مقصد (الزامی)
  - `arrival_date`: تاریخ ورود
  - `departure_date`: تاریخ خروج
  - `interests`: علایق کاربر مانند culture، food، nature، adventure، shopping، nightlife

- **محاسبه تعداد روزها:** با استفاده از `datetime` تعداد روزهای سفر را محاسبه کنید تا itinerary مناسب بسازید.

- **جستجوهای متنوع:** برای جمع‌آوری اطلاعات کامل، چندین query به پایگاه دانش بزنید:
  - جاذبه‌های اصلی و دیدنی‌ها
  - بهترین رستوران‌ها و غذاهای محلی
  - فعالیت‌ها و تفریحات
  - نواحی مناسب اقامت
  - (در صورت وجود interests) موضوعات خاص کاربر



- **مدیریت اطلاعات ناکافی:** اگر اطلاعات کافی در پایگاه دانش یافت نشد، از Tavily Search استفاده کنید.

- **مدیریت خطا:** خطاهای احتمالی خطای جستجو در پایگاه داده را مدیریت کنید.

- **استفاده از Decorator:** از `@tool` decorator برای تعریف این ابزار استفاده کنید.

**مثال درخواست کاربر:**
```
"Plan my trip to Paris from December 18th to 22th"
"Create a 4-day itinerary for Dubai with focus on food and culture"
```


</div>

In [ ]:
## WRITE YOUR CODE HERE

# <div style="text-align: center; direction: rtl; font-family: Vazir;"><h1 align="center" style="font-size: 24px; padding: 20px;">⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️<br>(۱۲.۵نمره) Creating Agent with LangGraph<br>⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️</h1></div>


<p dir="rtl" style="text-align: right; padding:30px; background-color:rgb(12, 12, 12); border-radius: 12px; color: white; font-family: Vazir;">
در این بخش، agent اصلی را با استفاده از  LangGraph و معماری ReAct  طراحی و پیاده‌سازی می‌کنیم.
برای اطلاعات بیشتر درباره معماری ReAct میتوانید مقاله زیر را مطالعه کنید:

https://arxiv.org/pdf/2210.03629

<p dir="rtl" style="padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
معماری Agent:<br>
• Reasoner Node: تصمیم‌گیری و استفاده از tools<br>
• Tools Node: اجرای tools<br>
• Workflow: مدیریت جریان مکالمه<br>
</p>
</p>


## <div style="text-align: center; direction: rtl; font-family: Vazir;">Defining System Message and Bind Tools</div>


<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">

یک System Message جامع بنویسید که:
- نقش agent را مشخص کند (Travel Assistant)
- لیست 8 tool موجود را معرفی کند
- دستورالعمل استفاده از هر tool را توضیح دهد
- سپس tools را به LLM bind کنید
</div>
<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
توجه کنید که مدیریت فرمت‌های تاریخ با توجه به ابزار مورد نظر را باید در  System Prompt در نظر بگیرید.
</div>

<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
پیام تایید bind موفق 7 tool
</p>

In [ ]:
# WRITE YOUR CODE HERE

## <div style="text-align: center; direction: rtl; font-family: Vazir;">Creating Graph Nodes</div>


<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
<p style="line-height: 1.8; text-align: right;">
Reasoner node را پیاده‌سازی کنید که:
- System message و تاریخچه مکالمه را ترکیب کند
- LLM را invoke کند
- پاسخ را برگرداند
</p>
</div>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
تعریف موفق reasoner node
</p>

In [ ]:
# WRITE YOUR CODE HERE

## <div style="text-align: center; direction: rtl; font-family: Vazir;">Creating and Compiling Graph</div>


<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
<p style="line-height: 1.8; text-align: right;">
StateGraph را بسازید و compile کنید و تصویر گراف را رسم کنید.

</p>
</div>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
• Graph compiled موفق<br>
• تصویر visualization گراف<br>
</p>

In [ ]:
# WRITE YOUR CODE HERE


## <div style="text-align: center; direction: rtl; font-family: Vazir;"> ساخت رابط کاربری CLI</div>


<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
<p style="line-height: 1.8; text-align: right;">
یک تابع chat_cli بنویسید که:
- در یک حلقه ورودی کاربر را بگیرد
- graph را invoke کند
- نام tool های استفاده شده را نمایش دهد
- پاسخ را چاپ کند
- دستورات exit، clear را پشتیبانی کند
</p>
</div>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
رابط CLI تعاملی که نمونه مکالمات را پشتیبانی کند
</p>

In [ ]:
# WRITE YOUR CODE HERE


# <div style="text-align: center; direction: rtl; font-family: Vazir;"><h1 align="center" style="font-size: 24px; padding: 20px;">⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️<br>(۱۰نمره)Testing Agent With Different Scenarios<br>⚜️━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━⚜️</h1></div>


<div dir="rtl" style="text-align: right; padding:10px; background-color:#6B7280;  border-radius: 12px; border: 2px solid rgb(2, 34, 22); font-family: Vazir;">
<p style="line-height: 1.8; text-align: right;">
چت بات خود را برای هرابزار با حداقل دو سناریو مختلف تست کنید


برای هر سناریو:
- Query کاربر را نمایش دهید
- Tool های استفاده شده را مشخص کنید
- خروجی نهایی را نمایش دهید
- کیفیت پاسخ را تحلیل کنید
</p>
</div>


<p dir='rtl' style="line-height: 2.0; text-align: right; font-family: Vazir; font-size: 16px; margin-top: 20px; color: white; background-color:rgb(0, 40, 30); padding: 30px; border-radius: 8px;">
🎯 <b>خروجی مورد انتظار:</b><br>
۱۴ تست موفق با خروجی کامل و تحلیل نتایج
</p>

In [ ]:
# WRITE YOUR CODE HERE


<div dir='rtl' style='background:#fffbe6; font-family: Vazir; border:1px dashed #f0ad4e; padding:12px; border-radius:8px; color:#111'>
✍️ <b>پاسخ تشریحی و تحلیل:</b><br>
<br>
<b>سناریو ۱:</b><br>
{{تحلیل نتیجه، tool های استفاده شده، کیفیت پاسخ}}<br>
<br>
<b>سناریو ۲:</b><br>
{{تحلیل نتیجه، tool های استفاده شده، کیفیت پاسخ}}<br>
<br>
<b>سناریو ۳:</b><br>
{{تحلیل نتیجه، tool های استفاده شده، کیفیت پاسخ}}<br>
<br>
<b>سناریو ۴:</b><br>
{{تحلیل نتیجه، tool های استفاده شده، کیفیت پاسخ}}<br>
<br>
<b>سناریو ۵:</b><br>
{{تحلیل نتیجه، tool های استفاده شده، کیفیت پاسخ}}<br>
<br>
<b>سناریو ۶:</b><br>
{{تحلیل نتیجه، tool های استفاده شده، کیفیت پاسخ}}<br>
<br>
<b>سناریو ۷:</b><br>
{{تحلیل نتیجه، tool های استفاده شده، کیفیت پاسخ}}<br>
<br>
<b>سناریو ۸:</b><br>
{{تحلیل نتیجه، tool های استفاده شده، کیفیت پاسخ}}<br>
<br>
<b>سناریو ۹:</b><br>
{{تحلیل نتیجه، tool های استفاده شده، کیفیت پاسخ}}<br>
<br>
<b>سناریو ۱۰:</b><br>
{{تحلیل نتیجه، tool های استفاده شده، کیفیت پاسخ}}<br>
<br>
<b>سناریو ۱۱:</b><br>
{{تحلیل نتیجه، tool های استفاده شده، کیفیت پاسخ}}<br>
<br>
<b>سناریو ۱۲:</b><br>
{{تحلیل نتیجه، tool های استفاده شده، کیفیت پاسخ}}<br>
<br>
<b>سناریو ۱۳:</b><br>
{{تحلیل نتیجه، tool های استفاده شده، کیفیت پاسخ}}<br>
<br>
<b>سناریو ۱۴:</b><br>
{{تحلیل نتیجه، tool های استفاده شده، کیفیت پاسخ}}<br>
<br>


<b>نتیجه‌گیری کلی:</b><br>
{{نقاط قوت و ضعف سیستم، پیشنهادات بهبود}}
</div>


# <h1 style="text-align: right;">**نکات مهم و قوانین تحویل**</h1>


<h4 dir="rtl" style="font-family: Vazir; width: 85%;">فایل ارسالی شما باید با فرمت زیر نامگذاری شود: <code>NLP_CA{n}_{LASTNAME}_{STUDENTID}.ipynb</code></h4>
<h4 dir="rtl" style="font-family: Vazir; width: 85%;">نحوه انجام تمرین:</h4>
<ul dir="rtl" style="font-family: Vazir; width: 85%; font-size: 16px;">
  <li>سلول‌های کد با برچسب <code>WRITE YOUR CODE HERE</code> را تکمیل کنید.</li>
  <li>برای پاسخ‌های متنی، متن <code>{{پاسخ_خود_را_اینجا_بنویسید}}</code> را با پاسخ خود جایگزین کنید.</li>
</ul>
<h4 dir="rtl" style="font-family: Vazir; width: 85%;">صداقت علمی:</h4> <ul dir="rtl" style="font-family: Vazir; width: 85%; font-size: 16px;"> <li>ما نوت‌بوک‌های تعداد مشخصی از دانشجویان که به صورت تصادفی انتخاب می‌شوند، بررسی خواهیم کرد. این بررسی‌ها اطمینان حاصل می‌کنند که کدی که نوشتید واقعاً پاسخ‌های موجود در نوت‌بوک شما را تولید می‌کند. اگر پاسخ‌های صحیح را در نوت‌بوک خود بدون کدی که واقعاً آن پاسخ‌ها را تولید کند تحویل دهید، این یک مورد جدی از عدم صداقت علمی محسوب می‌شود.</li> <li>ما همچنین بررسی‌های خودکاری را برای تشخیص سرقت علمی در نوت‌بوک‌های کولب انجام خواهیم داد. کپی کردن کد از دیگران نیز یک مورد جدی از عدم صداقت علمی محسوب می‌شود.</li> </ul>
<h4 dir="rtl" style="font-family: Vazir; width: 85%;">توضیحات تکمیلی:</h4> <ul dir="rtl" style="font-family: Vazir; width: 85%; font-size: 16px;">
<li>
خوانایی و دقت بررسی‌ها در گزارش نهایی از اهمیت ویژه‌ای برخوردار است. به تمرین‌هایی که به صورت کاغذی تحویل داده شوند یا به صورت عکس در سایت بارگذاری شوند، ترتیب اثری داده نخواهد شد.</li>
<li>
 همه‌ی کدهای پیوست گزارش بایستی قابلیت اجرای مجدد داشته باشند. در صورتی که برای اجرا مجدد آن‌ها نیاز به تنظیمات خاصی می‌باشد، بایستی تنظیمات مورد نیاز را نیز در گزارش خود ذکر کنید.  دقت کنید که  تمامی کدها باید توسط شما اجرا شده باشند و نتایج اجرا در فایل کدهای ارسالی مشخص باشد. به کدهایی که نتایج اجرای آن‌ها در فایل ارسالی مشخص نباشد نمره‌ای تعلق نمی‌گیرد.
</li>
<li>توجه کنید این تمرین باید به صورت تک‌نفره انجام شود و پاسخ‌های ارائه شده باید نتیجه فعالیت فرد نویسنده باشد (همفکری و به اتفاق هم نوشتن تمرین نیز ممنوع است). در صورت مشاهده
 تشابه به همه افراد مشارکت‌کننده، نمره تمرین صفر و به استاد گزارش می‌گردد.
 </li>

 <li>
لطفاً تمامی پاسخ‌های متنی خود را با <b>فونت وزیر (Vazir)</b> و به‌صورت <b>راست‌چین</b> بنویسید.  
از استفاده از فونت‌های پیش‌فرض خودداری کنید تا ظاهر نوت‌بوک شما یک‌دست و خوانا باشد.  
در بخش‌های تشریحی، سعی کنید پاسخ‌ها را کامل، منسجم و با رعایت نگارش فارسی بنویسید.  
همچنین، به چینش تمیز سلول‌ها و اجرای درست کدها توجه کنید تا تمرین شما با فرمت خواسته‌شده و استاندارد ارائه شود.
</li>
 <li>برای مطالعه بیشتر درباره‌ی فرمت Markdown می‌توانید از <a href="https://github.com/tajaddini/Persian-Markdown/blob/master/learn-MD.md">این لینک</a> مطالعه کنید.
 </li>
 </ul>
    

 </div>
